File: ios/LodoCore/Sources/LodoCore/MemorySearch.swift
import Foundation

/// 记忆功能的纯逻辑部分:文本截断、扩展名 → 类型映射、"问 AI"前的关键词粗排。
/// 放 LodoCore(而不是 app 层)是为了能在 swift test 里独立测试。
public enum MemorySearch {

    /// 提取文本存库与喂 AI 的统一截断上限(字符数)。
    public static let maxSourceChars = 8000
    /// "问 AI"时每条条目带入 prompt 的原文摘录上限。
    public static let maxExcerptChars = 400
    /// "问 AI"时最多带入 prompt 的条目数。
    public static let maxAskItems = 20

    /// 按字符数截断(Character 边界,emoji 不劈半);超长时结尾加省略号。
    public static func truncate(_ text: String, limit: Int = maxSourceChars) -> String {
        let trimmed = text.trimmingCharacters(in: .whitespacesAndNewlines)
        guard trimmed.count > limit else { return trimmed }
        return String(trimmed.prefix(limit)) + "…"
    }

    /// 文件扩展名 → 记忆类型;认不出的一律归 file(保留原文件不解析)。
    public static func kind(forExtension ext: String) -> MemoryKind {
        switch ext.lowercased() {
        case "pdf":
            return .pdf
        case "png", "jpg", "jpeg", "heic", "heif", "gif", "webp", "bmp", "tiff":
            return .image
        case "txt", "md", "markdown", "csv", "json", "log":
            return .text
        default:
            return .file
        }
    }

    /// "问 AI"前的本地粗排:按问题分词与条目文字的重合度计分取前 limit 条,
    /// 不足时按创建时间倒序补齐(保证 AI 总有语料可看)。
    /// 入参用元组而不是 MemoryItem,保持纯逻辑可测。
    public static func rank(
        question: String,
        items: [(index: Int, text: String, createdAt: Date)],
        limit: Int = maxAskItems
    ) -> [Int] {
        let terms = tokens(of: question)
        let scored = items.map { item -> (index: Int, score: Int, createdAt: Date) in
            let haystack = item.text.lowercased()
            let score = terms.reduce(0) { $0 + (haystack.contains($1) ? 1 : 0) }
            return (item.index, score, item.createdAt)
        }
        let matched = scored.filter { $0.score > 0 }
            .sorted { ($0.score, $0.createdAt) > ($1.score, $1.createdAt) }
        var result = matched.map(\.index)
        if result.count < limit {
            let rest = scored.filter { $0.score == 0 }
                .sorted { $0.createdAt > $1.createdAt }
                .map(\.index)
            result.append(contentsOf: rest)
        }
        return Array(result.prefix(limit))
    }

    /// "问 AI"时最多带入待办历史匹配结果的条数。
    public static let maxHistoryItems = 10

    /// 待办历史(已完成的 TaskItem)的关键词匹配:只返回真正命中的,不像 rank(:)
    /// 那样在命中不足时拿近期条目填充凑数——"上次做过 X"答不出来时不该被
    /// "最近随便一条待办"冒充答案,宁可返回空。
    public static func matchTaskHistory(
        question: String,
        items: [(index: Int, title: String)],
        limit: Int = maxHistoryItems
    ) -> [Int] {
        let terms = tokens(of: question)
        guard !terms.isEmpty else { return [] }
        let scored = items.map { item -> (index: Int, score: Int) in
            let haystack = item.title.lowercased()
            let score = terms.reduce(0) { $0 + (haystack.contains($1) ? 1 : 0) }
            return (item.index, score)
        }
        let matched = scored.filter { $0.score > 0 }.sorted { $0.score > $1.score }
        return Array(matched.prefix(limit).map(\.index))
    }

    /// 问题 → 检索词:按非字母数字切分取英文单词,中文再按双字滑窗补充
    /// (无分词库下对中文最实用的近似)。
    static func tokens(of question: String) -> [String] {
        let lowered = question.lowercased()
        var terms = lowered
            .components(separatedBy: CharacterSet.alphanumerics.inverted)
            .filter { $0.count >= 2 }
        let cjk = lowered.unicodeScalars.filter { $0.properties.isIdeographic }
        let chars = cjk.map(Character.init)
        if chars.count >= 2 {
            for i in 0..<(chars.count - 1) {
                terms.append(String(chars[i]) + String(chars[i + 1]))
            }
        } else if chars.count == 1 {
            terms.append(String(chars[0]))
        }
        return Array(Set(terms))
    }
}

File: android/app/src/main/java/com/lodo/app/core/MemorySearch.kt
package com.lodo.app.core

/**
 * 记忆检索的纯逻辑部分(截断/关键词打分/待办历史匹配),与 iOS LodoCore
 * MemorySearch.swift 对齐——iOS 还有基于端上 embedding 的语义检索,Android
 * 这一轮没有对应的端上 embedding 基础设施,退化成纯关键词版本;这与 iOS
 * "没有 embedding 时退化成关键词搜索"是同一个合法降级路径,不是残缺实现。
 * 不引 Android/Room 依赖,保持纯 Kotlin/JVM 可单测(与 Scheduler 同一个约定)。
 */
object MemorySearch {
    const val MAX_SOURCE_CHARS = 8000
    const val MAX_EXCERPT_CHARS = 400
    const val MAX_ASK_ITEMS = 20
    const val MAX_HISTORY_ITEMS = 10

    fun truncate(text: String, limit: Int = MAX_SOURCE_CHARS): String =
        if (text.length <= limit) text else text.take(limit)

    /** 按 token 重合数打分,分高优先、同分按新旧排;命中数不足 limit 时用最新
     * 条目补足,保证 AI 手头总有一些上下文(与 iOS MemorySearch.rank 一致)。 */
    fun <T> rank(
        question: String, items: List<T>, limit: Int,
        haystack: (T) -> String, createdAtMillis: (T) -> Long,
    ): List<T> {
        if (items.isEmpty()) return emptyList()
        val qTokens = tokens(question)
        val byRecency = items.sortedByDescending(createdAtMillis)
        if (qTokens.isEmpty()) return byRecency.take(limit)
        val scored = byRecency.map { item -> item to qTokens.count { tokens(haystack(item)).contains(it) } }
        val hit = scored.filter { it.second > 0 }
            .sortedWith(compareByDescending<Pair<T, Int>> { it.second }.thenByDescending { createdAtMillis(it.first) })
            .map { it.first }
        if (hit.size >= limit) return hit.take(limit)
        val padding = scored.filter { it.second == 0 }.map { it.first }.take(limit - hit.size)
        return hit + padding
    }

    /** 待办历史匹配:只返回真正命中关键词的,不做"补足"兜底——找不到就是
     * 没有,不能硬凑最近完成的事项充数,那会让"上次做过 X 吗"这类问题被
     * 误导性回答(与 iOS matchTaskHistory 一致)。 */
    fun <T> rankTaskHistory(
        question: String, items: List<T>, limit: Int, haystack: (T) -> String,
    ): List<T> {
        val qTokens = tokens(question)
        if (qTokens.isEmpty()) return emptyList()
        return items
            .map { it to qTokens.count { t -> tokens(haystack(it)).contains(t) } }
            .filter { it.second > 0 }
            .sortedByDescending { it.second }
            .take(limit)
            .map { it.first }
    }

    /** 字母数字词(≥2 字符)+ 中文按 2 字滑窗切分,与 iOS tokens(_:) 同一个思路。 */
    fun tokens(text: String): Set<String> {
        val lower = text.lowercase()
        val result = mutableSetOf<String>()
        Regex("[a-z0-9]{2,}").findAll(lower).forEach { result += it.value }
        Regex("[\\u4e00-\\u9fff]+").findAll(lower).forEach { run ->
            val s = run.value
            if (s.length == 1) {
                result += s
            } else {
                for (i in 0 until s.length - 1) result += s.substring(i, i + 2)
            }
        }
        return result
    }
}
